gateway.go

```diff
@@ -109,6 +113,62 @@ func (s *gatewayResource) gatewayHandler(w http.ResponseWriter, r *http.Request)
 	w.Write(packedResponse)
 }
 
+func (s *gatewayResource) marshalHandler(w http.ResponseWriter, r *http.Request) {
```

Comment:
With the new types of handlers, I would instead turn this into an `AppContentHandler` that returns the marshalled proto message. Maybe something like this?

```
type MarshalEchoAppHandler struct {
}

// Handle returns the input request as the response.
func (h MarshalEcho) Handle(binaryRequest []byte, metrics Metrics) ([]byte, error) {
        _, err := requestToProtoHTTP(parsedReq)
        if err != nil {
                return nil, err
        }
        response := &Response{}
        respEnc, err := proto.Marshal(response)
        if err != nil {
                return nil, err
        }
	metrics.Fire(metricsResultSuccess)
	return respEnc, nil
}
```

Reply:
thanks for the suggested refactoring! I will get back to it hopefully next week. Fir now I'm running load tests from the branch

Reply:
Hey Chris! I finally got back to this after error handling refactoring. I've introduced MarshalEncapsulationHandler and MarshalAppHandler as well as a lot of comments for this voodoo-magic code. Hopefully it made things clearer for the poor reader.